Issue 4531 - add bunny fonts

myroslavve opened the pull request:
# Description
In dashboard, added option to use bunny fonts instead of Google fonts:
![image](https://github.com/maxi-blocks/maxi-blocks/assets/53875005/17e45b99-0324-4d6b-99ff-83c418979d32)

<!---
Please include a summary of the changes and the related issue.
Please also include relevant motivation and context.
--->

Fixes #4531 

# How Has This Been Tested?
Checked in devtools that fonts load from fonts.bunny.net when enabled.
![image](https://github.com/maxi-blocks/maxi-blocks/assets/53875005/76086a64-4f5d-4d49-bec0-4f6a5be706e4)
Also checked that local fonts work with bunny fonts enabled.

<!---
Please describe the tests that you ran to verify your changes.
Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration.
--->

# Test checklist

<!--- Please remove the unnecessary checkbox --->

**_ Front/Back Testing _**

-   [ ] Test editor/frontend.
-   [ ] Test local fonts.

# Checklist

-   [ ] My code follows the style guidelines of this project
-   [ ] I have performed a self-review of my code
-   [ ] I have commented my code, particularly in hard-to-understand areas
-   [ ] My changes generate no new warnings/errors
-   [ ] I have added/updated tests that prove my fix is effective or that my feature works
-   [ ] New and existing unit tests pass locally with my changes


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a new setting to switch between Google Fonts and Bunny Fonts APIs for font loading.
  - Enhanced font loading behavior with improved URL construction based on font weights and styles.

- **Refactor**
  - Improved readability and maintainability of font-related methods and functions.

- **Bug Fixes**
  - Ensured font settings are updated correctly during installation.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

myroslavve committed:
Add bunny fonts option to the dashboard

myroslavve committed:
Implement frontend + local fonts for bunny fonts

myroslavve committed:
Implement bunny fonts on editor

myroslavve commented on core/admin/class-maxi-dashboard.php:
```diff
@@ -485,9 +485,14 @@ public function maxi_blocks_settings()
             $content .= $this->generate_item_header(__('Fonts and files', 'maxi-blocks'), false);
 
+
+            $description = '<h4>'.__('Use bunny.net fonts', 'maxi-blocks').'</h4>';
+            $description .= '<p>'.__('Use bunny.net CDN to serve fonts. GDPR compliant analogue to Google fonts', 'maxi-blocks').'</p>';
```
@ckp267 could you please check name/description for this in dashboard, maybe need to change or add anything.

elzadj commented:
I think we also need to update the list of fonts we are using (/src/extensions/text/fonts/fonts.js) - it's old, and we are missing some fonts.

kyrapieterse commented:
My chat with Svits - Let talk here

Why would you want local bunny fonts? Does it make it faster?

yes, local fonts are always faster
minus an external request
(ok, not always, if your server is shit, than external could be faster, but many people just don't want to depend on external services)

if google fonts goes down, half of internet shows Times New Roman

I agree if its working dont change it

elzadj commented:
Or we can just always download local fonts from bunny, then 3 options is ok.
Maybe we don't need to let people choose where to download them from. Again, need to check the legal staff.

ckp267 commented:
### Updated Font Settings Proposal for MaxiBlocks Dashboard 
( agreed in DM's with @myroslavve @elzadj @kyrapieterse )

_BEGIN - Settings accordion text and options_

### Fonts and files

**Use Bunny Fonts**
Privacy-friendly, GDPR compliant. Global CDN for fast loading. Wide selection of fonts available.
[x] Use Bunny Fonts (Default: On)

[x] Serve Bunny Fonts from CDN (Default: On)
Fastest option. Uses external CDN. No local storage required.

[ ] Serve Bunny Fonts locally (Default: Off)
Privacy-focused. May impact server performance. Requires local storage.

**Google Fonts**
[ ] Use Google Fonts (Default: Off)
Extensive font selection. Potential privacy concerns when using Google's CDN.

[x] Serve Google Fonts locally (Default: On if Google Fonts is used)
Blocks Google tracking. May impact server performance. Requires local storage.

[ ] Serve from Google CDN (Default: Off)
"Fastest option. Uses Google's CDN. Potential privacy (GDPR) implications."

**General note**
Local storage: Download and serve font files from your WordPress uploads directory. Improves privacy but may affect performance. Unused fonts are periodically removed to save space.

_END - Settings accordion text and options_

### Implementation Notes:
* Respect user's previous settings to avoid overwriting customizations
 - New users: bunny external by default (as per defaults above)
 - **IMPORTANT!!** Old users: whatever they used before. We need to test this and confirm!

* **IMPORTANT!!** Only ONE top level option. Either bunny OR google. Cannot have both at same time.

* Update the font list in /src/extensions/text/fonts/fonts.js.

* Refactor class-maxi-local-fonts.php for improved efficiency.

* Test current implementation periodic cleanup of unused local font files - make sure it works

myroslavve committed:
Update `fonts.js` with new fonts from google api

myroslavve committed:
Enable bunny fonts when plugin is installed

myroslavve committed:
Refactor local fonts

myroslavve commented:
@Olekrut 
Need to check 
>  - New users: bunny external by default (as per defaults above)
> - IMPORTANT!! Old users: whatever they used before. We need to test this and confirm!

So when you install plugin, even without opening maxi dashboard, frontend should have bunny fonts, and if you switch to this branch from master should still be google fonts.
Also, check fonts on frontend in general please, since the code was refactored.

myroslavve commented:
Also fixes #5430

Olekrut reviewed (changes requested):
https://jam.dev/c/b7ac4e6f-dca5-4062-be38-ee945f4daa97
1) created page on master > google fonts
2) switch brunch to this > fonts changed to bunny

elzadj commented:
> https://jam.dev/c/b7ac4e6f-dca5-4062-be38-ee945f4daa97
> 
> 1. created page on master > google fonts
> 2. switch brunch to this > fonts changed to bunny

@myroslavve could it be because the local env. already has bunny_fonts set to true? Need to figure out something here, at least for tests.

myroslavve commented:
> https://jam.dev/c/b7ac4e6f-dca5-4062-be38-ee945f4daa97
> 
> 1. created page on master > google fonts
> 2. switch brunch to this > fonts changed to bunny

If you have bunny fonts enabled before switching to master, it will remain enabled, and same for disabled state. So seems to be working right.

https://github.com/maxi-blocks/maxi-blocks/assets/53875005/edd19761-0b23-48c0-b40d-cb0d0653e0a3

elzadj reviewed (changes requested):
Great work! A couple of notes:
1. The font id on backed has spaces, if the font name has them. Ideally need to make it the same as on the frontend, like ```id='maxi-blocks-styles-font-roboto-css'``` 
![Screenshot from 2024-07-12 15-34-51](https://github.com/user-attachments/assets/3b011cba-d527-4fa7-9be2-03967239e546)
2. It's absolutely not clear that you can switch to Google fonts (or to Bunny fonts) here, or what you are using right now. Maybe just write it near the option? "You are using: Google fonts" or "You are using: Bunny fonts". @ckp267 
![screenshot-wordpress local-2024 07 12-15_49_15](https://github.com/user-attachments/assets/1a3c178d-6ec9-4baf-aea0-f8064dbbf013)
![screenshot-wordpress local-2024 07 12-16_01_07](https://github.com/user-attachments/assets/e1a20d91-3cdc-4f95-93c9-198edc9e65bd)
3. **Local bunny and google fonts don't work on frontend** - the style.css and font files are missing from uploads/maxi folder.
![screenshot-wordpress local-2024 07 12-16_07_52](https://github.com/user-attachments/assets/5422bdc5-993e-4780-8039-f930eb6c25a0)
@Olekrut please confirm, remove your wp-conent/uploads/maxi/fonts folder before activating the local option.
5. **Backend still loads the fonts from the external bunny site, if local option is enabled, is must be served from the local folder on backend too.**
6. Remove all commented out code, please.

Olekrut reviewed (changes requested):
hmm, yes local fonts are not working correct, only after changing font for some text and updating, then everything is appearing

myroslavve committed:
Fix local fonts not loading

myroslavve committed:
Remove unused code

myroslavve committed:
Enhance used font label in dashboard

myroslavve committed:
Merge branch 'master' into issue-4531

myroslavve commented:
> 1. The font id on backed has spaces, if the font name has them. Ideally need to make it the same as on the frontend, like `id='maxi-blocks-styles-font-roboto-css'`
>    ![Screenshot from 2024-07-12 15-34-51](https://private-user-images.githubusercontent.com/2401618/348263348-3b011cba-d527-4fa7-9be2-03967239e546.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwNTQ0MzMsIm5iZiI6MTcyMTA1NDEzMywicGF0aCI6Ii8yNDAxNjE4LzM0ODI2MzM0OC0zYjAxMWNiYS1kNTI3LTRmYTctOWJlMi0wMzk2NzIzOWU1NDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTVUMTQzNTMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTc0MDk2YzAyYzQ1NmE3NjZiNDlmMzkyMWFhYTY3MzI3MDc3ZGNjZTk0OTUxYTdiYzZjYmFiZWM2ZWMyZmE0NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.YCk3RAhu9s3FSWqOFXmLvTLjcBGRP0E3iCHshI9Yc1c)

Fixed

> 2. It's absolutely not clear that you can switch to Google fonts (or to Bunny fonts) here, or what you are using right now. Maybe just write it near the option? "You are using: Google fonts" or "You are using: Bunny fonts". @ckp267
>    ![screenshot-wordpress local-2024 07 12-15_49_15](https://private-user-images.githubusercontent.com/2401618/348266615-1a3c178d-6ec9-4baf-aea0-f8064dbbf013.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwNTQ0MzMsIm5iZiI6MTcyMTA1NDEzMywicGF0aCI6Ii8yNDAxNjE4LzM0ODI2NjYxNS0xYTNjMTc4ZC02ZWM5LTRiYWYtYWVhMC1mODA2NGRiYmYwMTMuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTVUMTQzNTMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTVlYWE1ZWM4ZGZlOWJjMjAzNmVhNGU5MjFiNzQyMjBjMzZhYTljNDU2MjBjODc3YTQ3ZTU4NWYwODcwOWI3NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.AnBSJHdTKUKEyfEo7vuEPtN5qyw9GinSB9aulHtH2vk)
>    ![screenshot-wordpress local-2024 07 12-16_01_07](https://private-user-images.githubusercontent.com/2401618/348269607-e1a20d91-3cdc-4f95-93c9-198edc9e65bd.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwNTQ0MzMsIm5iZiI6MTcyMTA1NDEzMywicGF0aCI6Ii8yNDAxNjE4LzM0ODI2OTYwNy1lMWEyMGQ5MS0zY2RjLTRmOTUtOTNjOS0xOThlZGM5ZTY1YmQuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTVUMTQzNTMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OThhZjUyOTQzZWRmNWVmMmRhZWMzZmNkZThiNTlhNDAzZDJmZmU5NWY5OWJmMjEwMzYzZTBhODlmOTE4N2M5ZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.PeGE1RT2rAOkEyS8I3xqR_9RU-6NapmAek5FbTV8dOU)

Added an indicator that shows which font provider is currently used, let me know if you have any other ideas to improve the labels
![image](https://github.com/user-attachments/assets/e6a52ba0-bca0-470b-88b8-036982f706a9)


> 3. **Local bunny and google fonts don't work on frontend** - the style.css and font files are missing from uploads/maxi folder.
>    ![screenshot-wordpress local-2024 07 12-16_07_52](https://private-user-images.githubusercontent.com/2401618/348271474-5422bdc5-993e-4780-8039-f930eb6c25a0.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwNTQ0MzMsIm5iZiI6MTcyMTA1NDEzMywicGF0aCI6Ii8yNDAxNjE4LzM0ODI3MTQ3NC01NDIyYmRjNS05OTNlLTQ3ODAtODAzOS1mOTMwZWI2YzI1YTAuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTVUMTQzNTMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGNmMGY3NmRkMDhmNTBhOGM0OWNlYzE4ZTRkZDAzODNlMjNhY2M1ZDc1ZGRmNjMzNjQ5M2M0MWU0YjVkMTI0NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.tTuyCPRC_4jrqoYejv-kdSgwODGO-oVZUYuXomRTQb0)

Fixed

> 4. **Backend still loads the fonts from the external bunny site, if local option is enabled, is must be served from the local folder on backend too.**

Doesn't work on master as well, I think it would be better to create a separate issue, because need to connect js and php code to save font files while in editor, could be complicated.

> 5. Remove all commented out code, please.

Done

elzadj reviewed (changes requested):
> > 1. The font id on backed has spaces, if the font name has them. Ideally need to make it the same as on the frontend, like `id='maxi-blocks-styles-font-roboto-css'`
> >    ![Screenshot from 2024-07-12 15-34-51](https://private-user-images.githubusercontent.com/2401618/348263348-3b011cba-d527-4fa7-9be2-03967239e546.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwNTQ0MzMsIm5iZiI6MTcyMTA1NDEzMywicGF0aCI6Ii8yNDAxNjE4LzM0ODI2MzM0OC0zYjAxMWNiYS1kNTI3LTRmYTctOWJlMi0wMzk2NzIzOWU1NDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTVUMTQzNTMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTc0MDk2YzAyYzQ1NmE3NjZiNDlmMzkyMWFhYTY3MzI3MDc3ZGNjZTk0OTUxYTdiYzZjYmFiZWM2ZWMyZmE0NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.YCk3RAhu9s3FSWqOFXmLvTLjcBGRP0E3iCHshI9Yc1c)
> 
> Fixed
> 

 1. I still see 'font-Roboto+500+normal' - please make it with 'maxi-blocks' to identify with our plugin, that's going to help with future debugging. Also please replace special symbols with '-' and make it all lower case. You can add 'backend' or 'editor' too, if you want.

2. I think we need to move it on top, looks weird now. Something like 
Fonts and files
You are using .... right now
Options
But it's a question for @ckp267 and @kyrapieterse 

4. Yeah, but on master we use ```new FontFace```, so the font file itself is not directly visible in the page source code. Now we see the external link right away if we check the source code. Need to do it asap, after this one, or go back to the FontFace API.

myroslavve committed:
Modify the font loader element's id

myroslavve commented:
> 1. I still see 'font-Roboto+500+normal' - please make it with 'maxi-blocks' to identify with our plugin, that's going to help with future debugging. Also please replace special symbols with '-' and make it all lower case. You can add 'backend' or 'editor' too, if you want.

Added 'maxi-blocks' to the id, not sure which special symbols to replace, since font names don't have any in them.

> 2. I think we need to move it on top, looks weird now. Something like
>    Fonts and files
>    You are using .... right now
>    Options
>    But it's a question for @ckp267 and @kyrapieterse

When it is on top it looks kind of weird, unless we add some styles
![image](https://github.com/user-attachments/assets/fa54b2a6-81c3-4fe8-b228-d8b102904b66)
or
![image](https://github.com/user-attachments/assets/cf73df52-0d8e-4f38-acfb-839a8f6bc626)
Looks a little better underneath the setting title:
![image](https://github.com/user-attachments/assets/2b27a614-546c-4e03-8971-1c5a183821ac)
so left it like this. Not sure if it is good enough though :shrug:

> 3. Yeah, but on master we use `new FontFace`, so the font file itself is not directly visible in the page source code. Now we see the external link right away if we check the source code. Need to do it asap, after this one, or go back to the FontFace API.

Okay, I will be working on the issue now, but in separate PR

myroslavve committed:
Update dashboard labels

kyrapieterse commented:
Looks a little better underneath the setting title: 

looks ok. MVP - when Its merged I can take another look

elzadj commented:
> Added 'maxi-blocks' to the id, not sure which special symbols to replace, since font names don't have any in them.

I meant "+" in the id. 

>ID and NAME tokens must begin with a letter ([A-Za-z]) and may be followed by any number of letters, digits ([0-9]), hyphens ("-"), underscores ("_"), colons (":"), and periods (".").

We can ignore it for now, and fix later, if we have any problems with it.
